pennylane_lightning_kokkos/_serialize.py

```diff
@@ -62,59 +71,107 @@ def _obs_has_kernel(obs: Observable) -> bool:
         ctype = np.complex128
-        obs_py = ObsStructKokkos_C128
+        rtype = np.int64
+        sparsehamiltonian_obs = SparseHamiltonianKokkos_C128
 
-    for o in tape.observables:
-        is_tensor = isinstance(o, Tensor)
+    spm = ob.sparse_matrix()
+    data = np.array(spm.data).astype(ctype)
+    indices = np.array(spm.indices).astype(rtype)
+    offsets = np.array(spm.indptr).astype(rtype)
 
-        wires = []
+    wires = []
+    wires_list = ob.wires.tolist()
+    wires.extend([wires_map[w] for w in wires_list])
 
-        if is_tensor:
-            for o_ in o.obs:
-                wires_list = o_.wires.tolist()
-                w = [wires_map[w] for w in wires_list]
-                wires.append(w)
-        else:
-            wires_list = o.wires.tolist()
-            w = [wires_map[w] for w in wires_list]
-            wires.append(w)
+    return sparsehamiltonian_obs(data, indices, offsets, wires)
 
-        name = o.name if is_tensor else [o.name]
 
-        params = []
+def _serialize_hermitian(ob, wires_map: dict, use_csingle: bool):
+    if use_csingle:
+        rtype = np.float32
+        hermitian_obs = HermitianObsKokkos_C64
+    else:
+        rtype = np.float64
+        hermitian_obs = HermitianObsKokkos_C128
+
+    data = qml.matrix(ob).astype(rtype).ravel(order="C")
+    return hermitian_obs(data, ob.wires.tolist())
+
+
+def _serialize_ob(ob, wires_map, use_csingle):
+    if isinstance(ob, Tensor):
+        return _serialize_tensor_ob(ob, wires_map, use_csingle)
+    elif ob.name == "Hamiltonian":
+        return _serialize_hamiltonian(ob, wires_map, use_csingle)
+    elif ob.name == "SparseHamiltonian":
+        return _serialize_sparsehamiltonian(ob, wires_map, use_csingle)
+    elif ob.name == "Hermitian":
+        raise TypeError(
+            f"Hermitian observables are not currently supported for adjoint differentiation. Please use Pauli-words only."
+        )
+    else:
+        return _serialize_named_ob(ob, wires_map, use_csingle)
 
-        if not _obs_has_kernel(o):
-            if is_tensor:
-                for o_ in o.obs:
-                    if not _obs_has_kernel(o_):
-                        params.append(qml.matrix(o_).ravel().astype(ctype))
-                    else:
-                        params.append([])
-            else:
-                params.append(qml.matrix(o).ravel().astype(ctype))
 
-        ob = obs_py(name, params, wires)
-        obs.append(ob)
+def _serialize_observables(tape: QuantumTape, wires_map: dict, use_csingle: bool = False) -> List:
+    """Serializes the observables of an input tape.
+    Args:
+        tape (QuantumTape): the input quantum tape
+        wires_map (dict): a dictionary mapping input wires to the device's backend wires
+        use_csingle (bool): whether to use np.complex64 instead of np.complex128
+    Returns:
+        list(ObservableKokkos_C64 or ObservableKokkos_C128): A list of observable objects compatible with the C++ backend
+    """
```

Comment:
Would be nice to have documentation like this in all functions.

Reply:
done.